File: swe.py
def encode_swe_literal(x: int) -> str:
    if x < 0:
        raise ValueError("SWE undefined for negative x")
    level = 0
    total = 0
    while True:
        count = 1 << level
        if x < total + count:
            index = x - total
            return format(index, f'0{level}b')
        total += count
        level += 1


def decode_swe_literal(bits: str) -> int:
    n = len(bits)
    base = sum(1 << i for i in range(n))
    suffix = int(bits, 2) if bits else 0
    return base + suffix


def encode_seed(value: int, arity: int = 1) -> str:
    if arity == 1:
        ab = "00"
    elif arity == 2:
        ab = "01"
    elif arity == 3:
        ab = "100"
    elif arity == 4:
        ab = "101"
    elif arity == 5:
        ab = "110"
    elif arity == "literal":
        ab = "111"
    else:
        raise ValueError("Arity must be 1–5 or 'literal'")
    # special case for zero
    if value == 0:
        return f"{ab}:00:0:0"
    payload = value + 1
    H4 = encode_swe_literal(payload)
    len4 = len(H4)
    H3 = encode_swe_literal(len4)
    len3 = len(H3)
    H2 = encode_swe_literal(len3)
    len2 = len(H2)
    H1 = encode_swe_literal(len2)
    return f"{ab}:{H1}:{H2}:{H3}:{H4}"


def decode_seed(code: str) -> tuple[int, int]:
    parts = code.split(":")
    ab = parts[0]
    if ab == "00":
        ar = 1
    elif ab == "01":
        ar = 2
    elif ab == "100":
        ar = 3
    elif ab == "101":
        ar = 4
    elif ab == "110":
        ar = 5
    elif ab == "111":
        ar = "literal"
    else:
        raise ValueError("Invalid arity")
    if parts[1:] == ["00", "0", "0"]:
        return 0, ar
    _, H1, H2, H3, H4 = parts
    p1 = decode_swe_literal(H4)
    return p1 - 1, ar

File: src/swe.rs
//! SWE seed decoding helpers
use crate::TelomereError;

fn decode_swe_literal(bits: &str) -> Result<usize, TelomereError> {
    let n = bits.len();
    let base = (1usize << n) - 1;
    let suffix = if bits.is_empty() {
        0
    } else {
        usize::from_str_radix(bits, 2)
            .map_err(|_| TelomereError::Header("invalid SWE literal".into()))?
    };
    Ok(base + suffix)
}

/// Decode a SWE seed string into a payload index and arity.
pub fn decode_seed(code: &str) -> Result<(usize, usize), TelomereError> {
    let parts: Vec<&str> = code.split(':').collect();
    if parts.is_empty() {
        return Err(TelomereError::Header("invalid seed".into()));
    }
    let arity = match parts[0] {
        "00" => 1usize,
        "01" => 2usize,
        "100" => 3usize,
        "101" => 4usize,
        "110" => 5usize,
        "111" => 0usize,
        _ => return Err(TelomereError::Header("invalid arity".into())),
    };
    if parts.len() >= 4 && parts[1] == "00" && parts[2] == "0" && parts[3] == "0" {
        return Ok((0, arity));
    }
    if parts.len() < 5 {
        return Err(TelomereError::Header("truncated seed".into()));
    }
    let payload = decode_swe_literal(parts[4])?;
    Ok((payload - 1, arity))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn decode_zero() {
        assert_eq!(decode_seed("00:00:0:0").unwrap(), (0, 1));
    }
}
